ts/main.ts

```diff
@@ -194,49 +263,187 @@ async function renderAllCardsInitial(): Promise<void> {
+    }
+    if (cardSort.filter.school) {
+      apiFilterUrl += `&school=${cardSort.filter.school}`;
+    }
+    const response = await fetch(
+      `https://www.dnd5eapi.co/api/spells${apiFilterUrl}`,
+    );
+
+    if (!response.ok) throw new Error(`Fetch error status: ${response.status}`);
+
+    const filteredSpellData = await response.json();
+
+    filteredArray = [];
+
+    // clears the filtered cards div
+    while ($spellsListFilteredCardsDiv.childNodes.length > 0) {
+      if (!$spellsListFilteredCardsDiv.firstChild) continue;
+      $spellsListFilteredCardsDiv.removeChild(
+        $spellsListFilteredCardsDiv.firstChild,
+      );
+    }
+
+    // renders cards into the filtered cards array
+    for (let i = 0; i < filteredSpellData.count; i++) {
+      const spellInfo = filteredSpellData.results[i];
+      filteredArray.push(
+        renderCard(spellInfo.name, spellInfo.level, spellInfo.url),
+      );
+    }
+    // appends cards into the filtered cards div
+    filteredArray.forEach((element) => {
+      $spellsListFilteredCardsDiv.appendChild(element);
+    });
+  } catch (err) {
+    console.error('Error:', err);
+  }
+}
+
+// resets the filter form when the 'reset filters' button is clicked,
+// then switches to the all cards view
+$clearFilterBtn.addEventListener('click', () => {
+  $spellsListFilterForm.reset();
+  $spellsListSearchInput.value = '';
+  cardSort.filter = {
+    name: '',
+    level: -1,
+    school: '',
+  };
+  $spellsListFilterDialog.close();
+  switchCardsList('all cards');
+});
+
+// resets the filter form to previous state when the 'cancel' button is clicked
+$cancelFilterBtn.addEventListener('click', () => {
+  $spellsListFilterDialog.close();
+  $spellsListFilterNameInput.value = cardSort.filter.name;
+  $spellsListFilterLevelSelect.value = cardSort.filter.level.toString();
+  $spellsListFilterSchoolSelect.value = cardSort.filter.school;
+});
+
+// event listener for filter spells form
+$spellsListFilterForm.addEventListener('submit', async (event: Event) => {
+  try {
+    event.preventDefault();
+
+    // store form values in the cardSort object
+    cardSort.filter.name = $spellsListFilterNameInput.value;
+    cardSort.filter.level = Number($spellsListFilterLevelSelect.value);
+    cardSort.filter.school = $spellsListFilterSchoolSelect.value;
+
+    $spellsListSearchInput.value = cardSort.filter.name;
+
+    $spellsListFilterDialog.close();
+
+    // switches to the all spells list if the filter form is empty,
+    // then returns from the function
+    if (
+      !cardSort.filter.name &&
+      cardSort.filter.level < 0 &&
+      !cardSort.filter.school
+    ) {
+      switchCardsList('all cards');
+      return;
+    }
+
+    // if the filter form is not empty, it will filter the spells and
+    // switch to the filtered list
+    await filterSpellsList();
+    switchCardsList('filtered cards');
+  } catch (err) {
+    console.error('Error:', err);
```

Comment:
Same thing here, can we just modify the elements we already have rather than deleting and recreating all of them?